File: Python/04-Strings/strings.py
"""
04 - Strings: the patterns that solve string problems, including KMP.

Run:  python strings.py
"""

from __future__ import annotations

import random
from collections import Counter, defaultdict


# ============================================================================
# 1. Reversal
# ============================================================================
def reverse_string(s: str) -> str:
    """O(n) time and space. Strings are immutable, so a copy is unavoidable."""
    return s[::-1]


def reverse_in_place(chars: list[str]) -> list[str]:
    """The interview version: a mutable list, two pointers, O(1) extra space."""
    lo, hi = 0, len(chars) - 1
    while lo < hi:
        chars[lo], chars[hi] = chars[hi], chars[lo]
        lo, hi = lo + 1, hi - 1
    return chars


# ============================================================================
# 2. Two pointers - palindromes
# ============================================================================
def is_palindrome(s: str) -> bool:
    """Ignores case and non-alphanumerics. O(n) time, O(1) space.

    Filtering inside the loop avoids building a cleaned copy of the string.
    """
    lo, hi = 0, len(s) - 1
    while lo < hi:
        while lo < hi and not s[lo].isalnum():
            lo += 1
        while lo < hi and not s[hi].isalnum():
            hi -= 1
        if s[lo].lower() != s[hi].lower():
            return False
        lo, hi = lo + 1, hi - 1
    return True


# ============================================================================
# 3. Frequency counting - anagrams
# ============================================================================
def are_anagrams(a: str, b: str) -> bool:
    """O(n) time, O(1) space (26 buckets). Sorting would be O(n log n)."""
    if len(a) != len(b):
        return False                      # cheap reject before any counting
    counts = [0] * 26
    for ca, cb in zip(a, b):
        counts[ord(ca) - ord("a")] += 1
        counts[ord(cb) - ord("a")] -= 1
    return all(c == 0 for c in counts)


def group_anagrams(words: list[str]) -> list[list[str]]:
    """Group words that are anagrams of each other. O(n * k) with k = word length.

    The key is the 26-length count tuple - hashable, and cheaper than sorting
    each word.
    """
    groups: dict[tuple, list[str]] = defaultdict(list)
    for word in words:
        counts = [0] * 26
        for ch in word:
            counts[ord(ch) - ord("a")] += 1
        groups[tuple(counts)].append(word)
    return list(groups.values())


def first_unique_char(s: str) -> int:
    """Index of the first non-repeating character, or -1. O(n) / O(1)."""
    freq = Counter(s)
    for i, ch in enumerate(s):
        if freq[ch] == 1:
            return i
    return -1


# ============================================================================
# 4. Sliding window
# ============================================================================
def longest_unique_substring(s: str) -> int:
    """Length of the longest substring without repeating characters. O(n).

    `left` never moves backwards, so both pointers together travel 2n steps.
    """
    last_seen: dict[str, int] = {}
    left = best = 0
    for right, ch in enumerate(s):
        if ch in last_seen and last_seen[ch] >= left:
            left = last_seen[ch] + 1
        last_seen[ch] = right
        best = max(best, right - left + 1)
    return best


def min_window_substring(s: str, t: str) -> str:
    """Smallest window of s containing every character of t (with multiplicity).

    O(n + m) time. Grow right until valid, then shrink left while it stays
    valid, recording the best window seen.
    """
    if not s or not t:
        return ""
    need = Counter(t)
    missing = len(t)                      # counts duplicates, unlike len(need)
    best = (float("inf"), 0, 0)
    left = 0

    for right, ch in enumerate(s):
        if need[ch] > 0:
            missing -= 1
        need[ch] -= 1                     # may go negative for surplus chars

        while missing == 0:               # window is valid: try to shrink it
            if right - left + 1 < best[0]:
                best = (right - left + 1, left, right)
            need[s[left]] += 1
            if need[s[left]] > 0:         # we just removed a needed character
                missing += 1
            left += 1

    return "" if best[0] == float("inf") else s[best[1]:best[2] + 1]


# ============================================================================
# 5. Building strings
# ============================================================================
def compress(s: str) -> str:
    """Run-length encoding: 'aabcccccaaa' -> 'a2b1c5a3'. O(n) time and space.

    Accumulate into a list and join once; `out += ...` would be O(n^2).
    """
    if not s:
        return ""
    parts: list[str] = []
    prev, count = s[0], 1
    for ch in s[1:]:
        if ch == prev:
            count += 1
        else:
            parts.append(f"{prev}{count}")
            prev, count = ch, 1
    parts.append(f"{prev}{count}")
    return "".join(parts)


# ============================================================================
# 6. Pattern matching
# ============================================================================
def naive_search(text: str, pattern: str) -> list[int]:
    """Every start position, checked character by character. O(n * m) worst case."""
    if not pattern:
        return []
    hits = []
    for i in range(len(text) - len(pattern) + 1):
        if text[i:i + len(pattern)] == pattern:
            hits.append(i)
    return hits


def build_lps(pattern: str) -> list[int]:
    """Longest proper Prefix that is also a Suffix, per position. O(m).

    lps[i] answers: if a mismatch happens after matching pattern[:i+1], how
    much of that match is still usable? That is what lets KMP avoid ever
    moving the text pointer backwards.
    """
    lps = [0] * len(pattern)
    length = 0                            # length of the current match
    i = 1
    while i < len(pattern):
        if pattern[i] == pattern[length]:
            length += 1
            lps[i] = length
            i += 1
        elif length:
            length = lps[length - 1]      # fall back, but do not advance i
        else:
            lps[i] = 0
            i += 1
    return lps


def kmp_search(text: str, pattern: str) -> list[int]:
    """All start indices of pattern in text. O(n + m) time, O(m) space."""
    if not pattern or len(pattern) > len(text):
        return []
    lps = build_lps(pattern)
    hits: list[int] = []
    i = j = 0                             # i indexes text, j indexes pattern
    while i < len(text):
        if text[i] == pattern[j]:
            i += 1
            j += 1
            if j == len(pattern):
                hits.append(i - j)
                j = lps[j - 1]            # keep going for overlapping matches
        elif j:
            j = lps[j - 1]                # reuse the known prefix
        else:
            i += 1
    return hits


def rabin_karp_search(text: str, pattern: str,
                      base: int = 256, modulus: int = 1_000_000_007) -> list[int]:
    """All start indices of pattern in text. O(n + m) expected, O(1) space.

    KMP avoids re-scanning by remembering prefix structure. Rabin-Karp takes a
    completely different route: HASH the pattern once, then slide a window over
    the text maintaining its hash in O(1) per step, and only compare characters
    when the hashes agree.

    THE ROLLING HASH. Treat the window as a base-B number modulo a prime:

        hash("abc") = (a * B^2 + b * B^1 + c * B^0) mod M

    Sliding one character right is three operations, not m:

        new = (old - leading * B^(m-1)) * B + trailing     (all mod M)

    Removing the leading digit is why B^(m-1) is precomputed - recomputing it
    inside the loop would make the whole thing O(n log m).

    THE VERIFICATION IS NOT OPTIONAL. Different strings can share a hash. On a
    match the characters must still be compared, or the function silently
    returns wrong positions. A hash equality is a CHEAP FILTER, never a proof:

        if window_hash == pattern_hash and text[i:i+m] == pattern:
                                           ^^^^^^^^^^^^^^^^^^^^^ this half

    Expected O(n + m); worst case O(n * m) if an adversary engineers collisions
    for every window. A large random-ish prime modulus makes that vanishingly
    unlikely in practice.

    WHY BOTHER when KMP is worst-case linear? Because the rolling hash
    generalises where KMP does not:
      - searching for MANY patterns at once (hash them all into a set)
      - 2-D pattern matching in a grid
      - "longest duplicate substring" - binary search the length, hash every
        window of that length, look for a repeat
      - detecting repeated blocks, plagiarism, rsync-style diffing
    """
    n, m = len(text), len(pattern)
    if not pattern or m > n:
        return []

    # B^(m-1) mod M - the weight of the character leaving the window.
    high_order = pow(base, m - 1, modulus)

    pattern_hash = 0
    window_hash = 0
    for i in range(m):                        # hash the pattern and the first window
        pattern_hash = (pattern_hash * base + ord(pattern[i])) % modulus
        window_hash = (window_hash * base + ord(text[i])) % modulus

    hits: list[int] = []
    for start in range(n - m + 1):
        # Hash equality is only a filter - the slice comparison is the proof.
        if window_hash == pattern_hash and text[start:start + m] == pattern:
            hits.append(start)

        if start + m < n:                     # roll the window one step right
            window_hash = (
                (window_hash - ord(text[start]) * high_order) * base
                + ord(text[start + m])
            ) % modulus
            # Python's % always returns a non-negative result, so no fix-up is
            # needed here. C, C++, Go and JavaScript all need one.

    return hits


def longest_duplicate_substring(s: str) -> str:
    """The longest substring appearing at least twice. O(n log n) expected.

    The payoff for having a rolling hash. The key observation is MONOTONICITY:
    if a duplicate of length L exists, so does one of every length below L
    (any prefix of it). That makes the answer binary-searchable.

    For each candidate length, hash every window of that length and look for a
    repeat - O(n) per check with a rolling hash, O(log n) checks.

    Hashes are stored with their positions so a collision can be resolved by
    comparing the actual substrings, keeping the result exact.
    """
    n = len(s)
    base, modulus = 256, 1_000_000_007

    def duplicate_of_length(length: int) -> str:
        if length == 0:
            return ""
        high_order = pow(base, length - 1, modulus)
        seen: dict[int, list[int]] = {}

        window_hash = 0
        for i in range(length):
            window_hash = (window_hash * base + ord(s[i])) % modulus

        for start in range(n - length + 1):
            for other in seen.get(window_hash, ()):     # verify, never trust
                if s[other:other + length] == s[start:start + length]:
                    return s[start:start + length]
            seen.setdefault(window_hash, []).append(start)

            if start + length < n:
                window_hash = (
                    (window_hash - ord(s[start]) * high_order) * base
                    + ord(s[start + length])
                ) % modulus
        return ""

    best = ""
    low, high = 1, n - 1
    while low <= high:                        # binary search on the LENGTH
        mid = (low + high) // 2
        found = duplicate_of_length(mid)
        if found:
            best = found
            low = mid + 1                     # try longer
        else:
            high = mid - 1                    # too long, try shorter
    return best


# ============================================================================
# 7. Everyday transformations
# ============================================================================
def reverse_words(sentence: str) -> str:
    """'the sky is blue' -> 'blue is sky the'. split() handles extra spaces."""
    return " ".join(reversed(sentence.split()))


def longest_common_prefix(words: list[str]) -> str:
    """O(total characters). Vertical scan, stops at the first mismatch."""
    if not words:
        return ""
    for i, ch in enumerate(words[0]):
        for other in words[1:]:
            if i >= len(other) or other[i] != ch:
                return words[0][:i]
    return words[0]


# ============================================================================
# demo
# ============================================================================
def demo() -> None:
    assert reverse_string("abc") == "cba"
    assert reverse_in_place(list("abcd")) == list("dcba")

    assert is_palindrome("A man, a plan, a canal: Panama")
    assert is_palindrome("")
    assert not is_palindrome("race a car")

    assert are_anagrams("listen", "silent")
    assert not are_anagrams("rat", "car")
    assert not are_anagrams("a", "ab")

    groups = group_anagrams(["eat", "tea", "tan", "ate", "nat", "bat"])
    assert sorted(len(g) for g in groups) == [1, 2, 3]

    assert first_unique_char("leetcode") == 0
    assert first_unique_char("aabb") == -1

    assert longest_unique_substring("abcabcbb") == 3
    assert longest_unique_substring("bbbbb") == 1
    assert longest_unique_substring("") == 0

    assert min_window_substring("ADOBECODEBANC", "ABC") == "BANC"
    assert min_window_substring("a", "aa") == ""

    assert compress("aabcccccaaa") == "a2b1c5a3"
    assert compress("") == ""

    text, pat = "ababcabcabababd", "ababd"
    assert naive_search(text, pat) == kmp_search(text, pat) == [10]
    assert build_lps("aabaaac") == [0, 1, 0, 1, 2, 2, 0]
    assert kmp_search("aaaa", "aa") == [0, 1, 2]     # overlapping matches
    assert kmp_search("abc", "") == []

    assert reverse_words("  the   sky is blue  ") == "blue is sky the"

    assert longest_common_prefix(["flower", "flow", "flight"]) == "fl"
    assert longest_common_prefix(["dog", "car"]) == ""

    # --- Rabin-Karp ----------------------------------------------------------
    assert rabin_karp_search("abracadabra", "abra") == [0, 7]
    assert rabin_karp_search("aaaa", "aa") == [0, 1, 2]     # overlapping
    assert rabin_karp_search("abc", "d") == []
    assert rabin_karp_search("abc", "") == []               # empty pattern
    assert rabin_karp_search("ab", "abc") == []             # pattern too long
    assert rabin_karp_search("aaa", "aaa") == [0]           # exact fit

    # Against naive search AND KMP, on a two-letter alphabet so that windows
    # collide constantly and the verification step actually gets exercised.
    random.seed(4)
    for _ in range(300):
        text = "".join(random.choice("ab") for _ in range(random.randint(0, 40)))
        pattern = "".join(random.choice("ab") for _ in range(random.randint(1, 5)))
        expected = naive_search(text, pattern)
        assert rabin_karp_search(text, pattern) == expected
        assert kmp_search(text, pattern) == expected

    # A tiny modulus forces genuine hash collisions - the verification step is
    # the only thing keeping the answer correct here.
    for _ in range(200):
        text = "".join(random.choice("abc") for _ in range(random.randint(0, 30)))
        pattern = "".join(random.choice("abc") for _ in range(random.randint(1, 4)))
        assert (rabin_karp_search(text, pattern, base=4, modulus=7)
                == naive_search(text, pattern))

    assert longest_duplicate_substring("banana") == "ana"
    assert longest_duplicate_substring("abcd") == ""        # nothing repeats
    assert longest_duplicate_substring("aaaa") == "aaa"
    assert longest_duplicate_substring("") == ""

    # Against an O(n^3) brute force over every pair of substrings.
    for _ in range(60):
        s = "".join(random.choice("abc") for _ in range(random.randint(0, 18)))
        expected_length = 0
        for length in range(1, len(s)):
            windows = [s[i:i + length] for i in range(len(s) - length + 1)]
            if len(set(windows)) < len(windows):
                expected_length = length
        assert len(longest_duplicate_substring(s)) == expected_length

    print("04-Strings (Python): all checks passed")
    print("  Rabin-Karp cross-checked against naive search and KMP, including")
    print("  with a deliberately tiny modulus that forces hash collisions")


if __name__ == "__main__":
    demo()

File: JavaScript/04-Strings/strings.js
/**
 * 04 - Strings: the patterns that solve string problems, including KMP.
 *
 * Run:  node strings.js
 */

import assert from "node:assert/strict";

// ============================================================================
// 1. Reversal
// ============================================================================
/** Spread iterates by CODE POINT, so this survives emoji. O(n) time and space. */
export function reverseString(s) {
  return [...s].reverse().join("");
}

/** The interview version: a mutable array of chars, two pointers, O(1) extra. */
export function reverseInPlace(chars) {
  let lo = 0;
  let hi = chars.length - 1;
  while (lo < hi) {
    [chars[lo], chars[hi]] = [chars[hi], chars[lo]];
    lo++;
    hi--;
  }
  return chars;
}

// ============================================================================
// 2. Two pointers - palindromes
// ============================================================================
/**
 * Ignores case and non-alphanumerics. O(n) time, O(1) space - filtering happens
 * inside the loop instead of allocating a cleaned copy.
 */
export function isPalindrome(s) {
  const isAlnum = (c) => /[a-z0-9]/i.test(c);
  let lo = 0;
  let hi = s.length - 1;
  while (lo < hi) {
    while (lo < hi && !isAlnum(s[lo])) lo++;
    while (lo < hi && !isAlnum(s[hi])) hi--;
    if (s[lo].toLowerCase() !== s[hi].toLowerCase()) return false;
    lo++;
    hi--;
  }
  return true;
}

// ============================================================================
// 3. Frequency counting - anagrams
// ============================================================================
const CODE_A = "a".charCodeAt(0);

/** O(n) time, O(1) space (26 buckets). Sorting would be O(n log n). */
export function areAnagrams(a, b) {
  if (a.length !== b.length) return false; // cheap reject
  const counts = new Array(26).fill(0);
  for (let i = 0; i < a.length; i++) {
    counts[a.charCodeAt(i) - CODE_A]++;
    counts[b.charCodeAt(i) - CODE_A]--;
  }
  return counts.every((c) => c === 0);
}

/** Group anagrams together. O(n * k) with k = word length. */
export function groupAnagrams(words) {
  const groups = new Map();
  for (const word of words) {
    const counts = new Array(26).fill(0);
    for (let i = 0; i < word.length; i++) counts[word.charCodeAt(i) - CODE_A]++;
    const key = counts.join("#"); // "#" separator: 1,11 must not equal 11,1
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(word);
  }
  return [...groups.values()];
}

/** Index of the first non-repeating character, or -1. O(n). */
export function firstUniqueChar(s) {
  const freq = new Map();
  for (const ch of s) freq.set(ch, (freq.get(ch) ?? 0) + 1);
  for (let i = 0; i < s.length; i++) if (freq.get(s[i]) === 1) return i;
  return -1;
}

// ============================================================================
// 4. Sliding window
// ============================================================================
/**
 * Longest substring with no repeated character. O(n).
 * `left` never moves backwards, so both pointers travel 2n steps in total.
 */
export function longestUniqueSubstring(s) {
  const lastSeen = new Map();
  let left = 0;
  let best = 0;
  for (let right = 0; right < s.length; right++) {
    const ch = s[right];
    if (lastSeen.has(ch) && lastSeen.get(ch) >= left) {
      left = lastSeen.get(ch) + 1; // jump past the previous occurrence
    }
    lastSeen.set(ch, right);
    best = Math.max(best, right - left + 1);
  }
  return best;
}

/**
 * Smallest window of s containing every character of t (with multiplicity).
 * O(n + m): grow right until valid, then shrink left while it stays valid.
 */
export function minWindowSubstring(s, t) {
  if (!s || !t) return "";
  const need = new Map();
  for (const ch of t) need.set(ch, (need.get(ch) ?? 0) + 1);

  let missing = t.length; // counts duplicates, unlike need.size
  let bestLen = Infinity;
  let bestStart = 0;
  let left = 0;

  for (let right = 0; right < s.length; right++) {
    const ch = s[right];
    if ((need.get(ch) ?? 0) > 0) missing--;
    need.set(ch, (need.get(ch) ?? 0) - 1); // may go negative for surplus

    while (missing === 0) {
      // valid window: shrink from the left
      if (right - left + 1 < bestLen) {
        bestLen = right - left + 1;
        bestStart = left;
      }
      const leaving = s[left];
      need.set(leaving, need.get(leaving) + 1);
      if (need.get(leaving) > 0) missing++; // we removed a needed character
      left++;
    }
  }
  return bestLen === Infinity ? "" : s.slice(bestStart, bestStart + bestLen);
}

// ============================================================================
// 5. Building strings
// ============================================================================
/** Run-length encoding: "aabcccccaaa" -> "a2b1c5a3". O(n) time and space. */
export function compress(s) {
  if (!s) return "";
  const parts = []; // push + join, never += in a loop
  let prev = s[0];
  let count = 1;
  for (let i = 1; i < s.length; i++) {
    if (s[i] === prev) {
      count++;
    } else {
      parts.push(prev, String(count));
      prev = s[i];
      count = 1;
    }
  }
  parts.push(prev, String(count));
  return parts.join("");
}

// ============================================================================
// 6. Pattern matching
// ============================================================================
/** Check every start position. O(n * m) worst case. */
export function naiveSearch(text, pattern) {
  const hits = [];
  if (!pattern || pattern.length > text.length) return hits;
  for (let i = 0; i + pattern.length <= text.length; i++) {
    let j = 0;
    while (j < pattern.length && text[i + j] === pattern[j]) j++;
    if (j === pattern.length) hits.push(i);
  }
  return hits;
}

/**
 * lps[i] = length of the longest proper prefix of pattern[0..i] that is also a
 * suffix of it. On a mismatch it says how much of the match is still usable,
 * which is why KMP never moves the text pointer backwards. O(m).
 */
export function buildLPS(pattern) {
  const lps = new Array(pattern.length).fill(0);
  let len = 0;
  let i = 1;
  while (i < pattern.length) {
    if (pattern[i] === pattern[len]) {
      lps[i++] = ++len;
    } else if (len > 0) {
      len = lps[len - 1]; // fall back, do NOT advance i
    } else {
      lps[i++] = 0;
    }
  }
  return lps;
}

/** All start indices of pattern in text. O(n + m) time, O(m) space. */
export function kmpSearch(text, pattern) {
  const hits = [];
  if (!pattern || pattern.length > text.length) return hits;
  const lps = buildLPS(pattern);
  let i = 0;
  let j = 0;
  while (i < text.length) {
    if (text[i] === pattern[j]) {
      i++;
      j++;
      if (j === pattern.length) {
        hits.push(i - j);
        j = lps[j - 1]; // keep going: overlapping matches count
      }
    } else if (j > 0) {
      j = lps[j - 1];
    } else {
      i++;
    }
  }
  return hits;
}

/**
 * All start indices of `pattern` in `text`. O(n + m) expected, O(1) space.
 *
 * KMP avoids re-scanning by remembering prefix structure. Rabin-Karp takes a
 * different route: HASH the pattern once, slide a window over the text keeping
 * its hash in `O(1)` per step, and only compare characters when hashes agree.
 *
 * **The rolling hash.** Treat the window as a base-B number modulo a prime:
 *
 *     hash("abc") = (a * B^2 + b * B^1 + c * B^0) mod M
 *
 * Sliding one character right is three operations, not m:
 *
 *     new = (old - leading * B^(m-1)) * B + trailing        (all mod M)
 *
 * Removing the leading digit is why `B^(m-1)` is precomputed - recomputing it
 * inside the loop would make the whole thing `O(n log m)`.
 *
 * **The verification is not optional.** Different strings can share a hash. On
 * a hash match the characters must still be compared, or the function silently
 * returns wrong positions. Hash equality is a CHEAP FILTER, never a proof.
 *
 * **JavaScript specific, twice over:**
 *   - `%` on a negative left operand returns a negative result (it is
 *     remainder, not modulo), so the subtraction needs a `+ modulus` fix-up.
 *   - `hash * base` with a 1e9 modulus reaches ~2.6e11, and rolling that into
 *     the next step exceeds `Number.MAX_SAFE_INTEGER`. **BigInt** keeps it
 *     exact. A smaller modulus would fit in a Number but collide far more.
 *
 * Expected `O(n + m)`; worst case `O(n * m)` if an adversary engineers
 * collisions. Worth it over KMP because the rolling hash generalises: many
 * patterns at once, 2-D grid matching, longest duplicate substring, rsync.
 */
export function rabinKarpSearch(text, pattern, base = 256n, modulus = 1000000007n) {
  const n = text.length;
  const m = pattern.length;
  if (m === 0 || m > n) return [];

  base = BigInt(base);
  modulus = BigInt(modulus);

  // B^(m-1) mod M - the weight of the character leaving the window.
  let highOrder = 1n;
  for (let i = 0; i < m - 1; i++) highOrder = (highOrder * base) % modulus;

  let patternHash = 0n;
  let windowHash = 0n;
  for (let i = 0; i < m; i++) {
    patternHash = (patternHash * base + BigInt(pattern.charCodeAt(i))) % modulus;
    windowHash = (windowHash * base + BigInt(text.charCodeAt(i))) % modulus;
  }

  const hits = [];
  for (let start = 0; start + m <= n; start++) {
    // Hash equality is only a filter - the slice comparison is the proof.
    if (windowHash === patternHash && text.slice(start, start + m) === pattern) {
      hits.push(start);
    }

    if (start + m < n) {
      // roll the window one step right
      const leaving = (BigInt(text.charCodeAt(start)) * highOrder) % modulus;
      windowHash = (windowHash - leaving) % modulus;
      if (windowHash < 0n) windowHash += modulus; // % is remainder, not modulo
      windowHash = (windowHash * base + BigInt(text.charCodeAt(start + m))) % modulus;
    }
  }
  return hits;
}

/**
 * The longest substring appearing at least twice. O(n log n) expected.
 *
 * The payoff for having a rolling hash. The key observation is MONOTONICITY:
 * if a duplicate of length L exists, so does one of every shorter length (any
 * prefix of it). That makes the answer binary-searchable.
 *
 * For each candidate length, hash every window and look for a repeat - `O(n)`
 * per check with a rolling hash, `O(log n)` checks.
 *
 * Hashes are stored with their positions so a collision is resolved by
 * comparing the real substrings, keeping the result exact.
 */
export function longestDuplicateSubstring(s) {
  const n = s.length;
  const base = 256n;
  const modulus = 1000000007n;

  const duplicateOfLength = (length) => {
    if (length === 0) return "";
    let highOrder = 1n;
    for (let i = 0; i < length - 1; i++) highOrder = (highOrder * base) % modulus;

    const seen = new Map();
    let windowHash = 0n;
    for (let i = 0; i < length; i++) {
      windowHash = (windowHash * base + BigInt(s.charCodeAt(i))) % modulus;
    }

    for (let start = 0; start + length <= n; start++) {
      for (const other of seen.get(windowHash) ?? []) {
        // verify, never trust
        if (s.slice(other, other + length) === s.slice(start, start + length)) {
          return s.slice(start, start + length);
        }
      }
      if (!seen.has(windowHash)) seen.set(windowHash, []);
      seen.get(windowHash).push(start);

      if (start + length < n) {
        const leaving = (BigInt(s.charCodeAt(start)) * highOrder) % modulus;
        windowHash = (windowHash - leaving) % modulus;
        if (windowHash < 0n) windowHash += modulus;
        windowHash =
          (windowHash * base + BigInt(s.charCodeAt(start + length))) % modulus;
      }
    }
    return "";
  };

  let best = "";
  let low = 1;
  let high = n - 1;
  while (low <= high) {
    // binary search on the LENGTH
    const mid = Math.floor((low + high) / 2);
    const found = duplicateOfLength(mid);
    if (found) {
      best = found;
      low = mid + 1; // try longer
    } else {
      high = mid - 1; // too long, try shorter
    }
  }
  return best;
}

// ============================================================================
// 7. Everyday transformations
// ============================================================================
/** split(/\s+/) collapses runs of whitespace; filter drops empties. */
export function reverseWords(sentence) {
  return sentence.trim().split(/\s+/).reverse().join(" ");
}

/** Vertical scan: compare column by column. O(total characters). */
export function longestCommonPrefix(words) {
  if (words.length === 0) return "";
  for (let i = 0; i < words[0].length; i++) {
    for (let w = 1; w < words.length; w++) {
      if (i >= words[w].length || words[w][i] !== words[0][i]) {
        return words[0].slice(0, i);
      }
    }
  }
  return words[0];
}

// ============================================================================
// demo
// ============================================================================
function demo() {
  assert.equal(reverseString("abc"), "cba");
  assert.equal(reverseString("ab😀"), "😀ba"); // survives surrogate pairs
  assert.deepEqual(reverseInPlace([..."abcd"]), [..."dcba"]);

  assert.ok(isPalindrome("A man, a plan, a canal: Panama"));
  assert.ok(isPalindrome(""));
  assert.ok(!isPalindrome("race a car"));

  assert.ok(areAnagrams("listen", "silent"));
  assert.ok(!areAnagrams("rat", "car"));
  assert.ok(!areAnagrams("a", "ab"));

  const groups = groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]);
  assert.deepEqual(
    groups.map((g) => g.length).sort(),
    [1, 2, 3],
  );

  assert.equal(firstUniqueChar("leetcode"), 0);
  assert.equal(firstUniqueChar("aabb"), -1);

  assert.equal(longestUniqueSubstring("abcabcbb"), 3);
  assert.equal(longestUniqueSubstring("bbbbb"), 1);
  assert.equal(longestUniqueSubstring(""), 0);

  assert.equal(minWindowSubstring("ADOBECODEBANC", "ABC"), "BANC");
  assert.equal(minWindowSubstring("a", "aa"), "");

  assert.equal(compress("aabcccccaaa"), "a2b1c5a3");
  assert.equal(compress(""), "");

  const text = "ababcabcabababd";
  const pat = "ababd";
  assert.deepEqual(naiveSearch(text, pat), [10]);
  assert.deepEqual(kmpSearch(text, pat), [10]);
  assert.deepEqual(buildLPS("aabaaac"), [0, 1, 0, 1, 2, 2, 0]);
  assert.deepEqual(kmpSearch("aaaa", "aa"), [0, 1, 2]); // overlapping
  assert.deepEqual(kmpSearch("abc", ""), []);

  assert.equal(reverseWords("  the   sky is blue  "), "blue is sky the");

  assert.equal(longestCommonPrefix(["flower", "flow", "flight"]), "fl");
  assert.equal(longestCommonPrefix(["dog", "car"]), "");

  // --- Rabin-Karp -----------------------------------------------------------
  assert.deepEqual(rabinKarpSearch("abracadabra", "abra"), [0, 7]);
  assert.deepEqual(rabinKarpSearch("aaaa", "aa"), [0, 1, 2]); // overlapping
  assert.deepEqual(rabinKarpSearch("abc", "d"), []);
  assert.deepEqual(rabinKarpSearch("abc", ""), []); // empty pattern
  assert.deepEqual(rabinKarpSearch("ab", "abc"), []); // pattern too long
  assert.deepEqual(rabinKarpSearch("aaa", "aaa"), [0]); // exact fit

  // Deterministic PRNG so a failure is always reproducible.
  let seed = 4;
  const random = () => {
    seed = (seed * 1103515245 + 12345) & 0x7fffffff;
    return seed / 0x7fffffff;
  };
  const randInt = (lo, hi) => lo + Math.floor(random() * (hi - lo + 1));
  const randomString = (length, alphabet) =>
    Array.from({ length }, () => alphabet[randInt(0, alphabet.length - 1)]).join("");

  // Against naive search AND KMP, on a two-letter alphabet so that windows
  // collide constantly and the verification step actually gets exercised.
  for (let trial = 0; trial < 300; trial++) {
    const text = randomString(randInt(0, 40), "ab");
    const pattern = randomString(randInt(1, 5), "ab");
    const expected = naiveSearch(text, pattern);
    assert.deepEqual(rabinKarpSearch(text, pattern), expected);
    assert.deepEqual(kmpSearch(text, pattern), expected);
  }

  // A tiny modulus forces genuine hash collisions - the verification step is
  // the only thing keeping the answer correct here.
  for (let trial = 0; trial < 200; trial++) {
    const text = randomString(randInt(0, 30), "abc");
    const pattern = randomString(randInt(1, 4), "abc");
    assert.deepEqual(
      rabinKarpSearch(text, pattern, 4n, 7n),
      naiveSearch(text, pattern),
    );
  }

  assert.equal(longestDuplicateSubstring("banana"), "ana");
  assert.equal(longestDuplicateSubstring("abcd"), ""); // nothing repeats
  assert.equal(longestDuplicateSubstring("aaaa"), "aaa");
  assert.equal(longestDuplicateSubstring(""), "");

  // Against an O(n^3) brute force over every pair of substrings.
  for (let trial = 0; trial < 60; trial++) {
    const s = randomString(randInt(0, 18), "abc");
    let expectedLength = 0;
    for (let length = 1; length < s.length; length++) {
      const windows = [];
      for (let i = 0; i + length <= s.length; i++) windows.push(s.slice(i, i + length));
      if (new Set(windows).size < windows.length) expectedLength = length;
    }
    assert.equal(longestDuplicateSubstring(s).length, expectedLength);
  }

  console.log("04-Strings (JavaScript): all checks passed");
  console.log(
    "  Rabin-Karp cross-checked against naive search and KMP, including\n" +
      "  with a deliberately tiny modulus that forces hash collisions",
  );
}

demo();
